File: static/scripts/Subfamilia.js
function filter(page = 1) {
    var IdSubFamilia = document.getElementById('IdSubFamilia').value;
    var Descripcion = document.getElementById('Descripcion').value;
    var tbody = document.getElementById('tableBody');
    var paginationDiv = document.getElementById('paginationDiv');
    var oldPagination = document.getElementById('pagination');
    if (oldPagination) {
        paginationDiv.removeChild(oldPagination);
    }
    var loading = document.getElementById('loading');
    loading.style.display = 'flex';

    fetch(`/search_subfamilia?page=${page}`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            IdSubFamilia: IdSubFamilia,
            Descripcion: Descripcion
        })
    })
        .then(response => response.json())
        .then(data => {
            while (tbody.firstChild) {
                tbody.removeChild(tbody.firstChild);
            }
            data.results.forEach((row, index) => {
                var tr = document.createElement('tr');
                tr.setAttribute("data-index", index);
                tr.setAttribute('data-id', row.IdSubFamilia);
                tr.setAttribute('data-d', row.Descripcion);
                tr.setAttribute('data-f', row.IdFamilia);
                tr.setAttribute('data-s', row.Servicio);
                tr.innerHTML = `
                    <td>${row.IdSubFamilia}</td>
                    <td>${row.Descripcion}</td>
                    <td>${row.IdFamilia}</td>
                    <td>${row.Servicio}</td>
            
                    <td><a href="#" title="Editar" onclick="editRow(${index})"><i class="fas fa-pencil-alt"></i></a>
                    <a href="#" title="Excluir linha" onclick="deleterow(${index})"><i class="fas fa-trash-alt"></i></a></td>
           
                `;
                tbody.appendChild(tr);
                console.log(row);

            });
            var paginationDiv = document.getElementById('paginationDiv');
            if (paginationDiv.firstChild) {
                paginationDiv.removeChild(paginationDiv.firstChild);
            }

            if (data.num_pages > 1) {
                var pagination = document.createElement('select');
                pagination.id = 'pagination';
                for (var i = 1; i <= data.num_pages; i++) {
                    var option = document.createElement('option');
                    option.value = i;
                    option.text = `Pagina ${i}`;
                    if (i == data.page) {
                        option.selected = true;
                    }
                    pagination.appendChild(option);
                }
                pagination.addEventListener('change', event => {
                    filter(event.target.value);
                });
                paginationDiv.appendChild(pagination);
            }
            loading.style.display = 'none';
        })
        .catch(error => {
            console.error(error);
        });
}
filter();

function showConfig() {
    var popup = document.getElementById("configPopup");
    popup.style.display = "block";
}

function hideConfig() {
    var popup = document.getElementById("configPopup");
    popup.style.display = "none";
}

function GudarConfig() {
    var IdSubFamilia = document.getElementById("configIdSubFamilia").value;
    var Descripcion = document.getElementById("configDescripcion").value;
    var IdFamilia = document.getElementById("configIdFamilia").value;
    var Servicio = document.getElementById("configIdServicio").value;
    
    fetch("/insert_subfamilia", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            IdSubFamilia: IdSubFamilia,
            Descripcion: Descripcion,
            IdFamilia: IdFamilia,
            Servicio: Servicio,
            
           
        }),
    })
        .then((response) => {
            if (response.status === 200) {
                alert("Dados inseridos com sucesso!");
            } else {
                throw new Error("Erro ao inserir os dados. " );
            }
            hideConfig();
        })
        .catch((error) => {
            console.error(error);
            alert("Erro ao inserir os dados." + error.message);
            console.log({ "IdSubFamilia": IdSubFamilia, "Descripcion": Descripcion, "IdFamilia": IdFamilia, "configIdServicio": Servicio});
            hideConfig();
        });
}

function showEdit() {
    var popup = document.getElementById("editPopup");
    popup.style.display = "block";
}

function editRow(index) {
    var row = document.querySelector(`#tableBody tr[data-index='${index}']`);
    if (row) {
        var IdSubFamilia = row.querySelector('td:nth-child(1)').innerText;
        var Descripcion = row.querySelector('td:nth-child(2)').innerText;
        var IdFamilia = row.querySelector('td:nth-child(3)').innerText;
        var Servicio = row.querySelector('td:nth-child(4)').innerText;
        
        document.getElementById("editIdSubFamilia").value = IdSubFamilia;
        document.getElementById("editDescripcion").value = Descripcion;
        document.getElementById("editIdFamilia").value = IdFamilia;
        document.getElementById("editIdServicio").value = Servicio;
        
        document.getElementById("editIdSubFamilia").setAttribute("data-old-value", IdSubFamilia);
        document.getElementById("editDescripcion").setAttribute("data-old-value", Descripcion);
        document.getElementById("editIdFamilia").setAttribute("data-old-value", IdFamilia);
        document.getElementById("editIdServicio").setAttribute("data-old-value", Servicio);
        
        showEdit();
    } else {
        alert(`Linha com ID ${index} não encontrada.`);
        console.error(`Linha com ID ${index} não encontrada.`);
        return;
    }
}

function hideEdit() {
    var popup = document.getElementById("editPopup");
    popup.style.display = "none";
}

function updateRow() {
    var oldIdSubFamilia = document.getElementById("editIdSubFamilia").getAttribute("data-old-value");
    var oldDescripcion = document.getElementById("editDescripcion").getAttribute("data-old-value");
    var oldIdFamilia = document.getElementById("editIdFamilia").getAttribute("data-old-value");
    var oldServicio = document.getElementById("editIdServicio").getAttribute("data-old-value");
    


    var newIdSubFamilia = document.getElementById("editIdSubFamilia").value;
    var newDescripcion = document.getElementById("editDescripcion").value;
    var newIdFamilia = document.getElementById("editIdFamilia").value;
    var newServicio = document.getElementById("editIdServicio").value;
    console.log(oldServicio,newServicio)
  
    fetch("/update_subfamilia", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            oldIdSubFamilia: oldIdSubFamilia,
            oldDescripcion: oldDescripcion,
            oldIdFamilia: oldIdFamilia,
            oldServicio: oldServicio,
            
            newIdSubFamilia: newIdSubFamilia,
            newDescripcion: newDescripcion,
            newIdFamilia: newIdFamilia,
            newServicio: newServicio,
        }),
    })
        .then((response) => {
            if (response.status === 200) {
                alert("Dados atualizados com sucesso!");
                filter();
            } else {
                alert("Erro ao atualizar os dados.");
            }
            hideEdit();
        });
}
function deleterow(index) {
    var row = document.querySelector(`#tableBody tr[data-index='${index}']`);
    var delIdSubFamilia = row.querySelector('td:nth-child(1)').innerText;
    var delDescripcion = row.querySelector('td:nth-child(2)').innerText;
    var delIdFamilia = row.querySelector('td:nth-child(3)').innerText;
    var delServicio = row.querySelector('td:nth-child(4)').innerText;
    
    if (confirm("¿Estás seguro de que deseas eliminar esta fila?")) {
        fetch("/delete_subfamilia", {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                delIdSubFamilia: delIdSubFamilia,
                delDescripcion: delDescripcion,
                delIdFamilia: delIdFamilia,
                delServicio: delServicio,
                
                
            }),
        })
            .then((response) => {
                if (response.status === 200) {
                    alert("Fila borrada con éxito!");
                    filter();
                } else {
                    alert("Erro al borrar la fila.");
                }
                hideEdit();
            });
    }
}

File: python/Subfamilia.py
import pyodbc
from flask import Flask, request, jsonify

import math
from conn import SubFamilia,Session
from sqlalchemy import and_,text


app = Flask(__name__, template_folder='templates')

def search_subfamilia(IdSubFamilia, Descripcion, page):
    session = Session()
    registros_por_pagina = 50
    primeiro_registro = (page - 1) * registros_por_pagina

    query = session.query(SubFamilia).filter(
        and_(
            (SubFamilia.IdSubFamilia == IdSubFamilia) if IdSubFamilia else True,
            SubFamilia.Descripcion.contains(Descripcion) if Descripcion else True
        )
    )

    total_registros = query.count()
    num_paginas = math.ceil(total_registros / registros_por_pagina)
    results = query.order_by(SubFamilia.IdSubFamilia).offset(primeiro_registro).limit(registros_por_pagina).all()
    print(query)
    session.close()

    return results, num_paginas

@app.route('/search_subfamilia')
def search_subfamilia_route():
    data = request.get_json()
    IdSubFamilia = data.get('IdSubFamilia')
    Descripcion = data.get('Descripcion')
    page = int(request.args.get('page', '1'))

    results, num_paginas = search_subfamilia(IdSubFamilia, Descripcion, page)
    
    dict_results = [
        {
            'IdSubFamilia': row.IdSubFamilia,
            'Descripcion': row.Descripcion,
            'IdFamilia': row.IdFamilia,
            'Servicio': row.Servicio,
            'Activo': row.Activo
        }
        for row in results
    ]

    return jsonify({'results': dict_results, 'page': page, 'num_pages': num_paginas})


@app.route('/insert_subfamilia', methods=['POST'])
def insert_subfamilia():
    data = request.get_json()
    IdSubFamilia = data['IdSubFamilia']
    Descripcion = data['Descripcion']
    IdFamilia = data['IdFamilia']
    Servicio = data['Servicio']
    Activo = 1

    try:
        session = Session()
        new_subfamilia = SubFamilia(IdSubFamilia, Descripcion, IdFamilia,Servicio,Activo)
        session.add(new_subfamilia)
        session.commit()
        session.close()

        return jsonify({'success': True}), 200

    except pyodbc.Error as e:
        print(e)
        return jsonify({'success': False, 'error': str(e)}), 500


@app.route('/update_subfamilia', methods=['POST'])
def update_subfamilia():
    data = request.get_json()
    newIdSubFamilia = data['newIdSubFamilia']
    newDescripcion = data['newDescripcion']
    newIdFamilia = data['newIdFamilia']
    newServicio = data['newServicio']
    oldIdSubFamilia = data['oldIdSubFamilia']
    oldDescripcion = data['oldDescripcion']
    oldIdFamilia = data['oldIdFamilia']
    oldServicio = data['oldServicio']
    Activo = 1

    try:
    
        session = Session()
        Subfamilia = session.query(SubFamilia).filter(
            and_(
                SubFamilia.IdSubFamilia == oldIdSubFamilia,
                text("convert(varchar, Descripcion) = :desc").params(desc=oldDescripcion),
                SubFamilia.IdFamilia == oldIdFamilia,
                SubFamilia.Servicio == oldServicio
               
            )
        ).first()
   

        if Subfamilia:
            Subfamilia.IdSubFamilia = newIdSubFamilia
            Subfamilia.Descripcion = newDescripcion
            Subfamilia.IdFamilia = newIdFamilia
            Subfamilia.Servicio = newServicio
            Subfamilia.Activo = Activo 
            session.commit()
            session.close()

            return jsonify({'success': True}), 200
        else:
            return jsonify({'success': False, 'error': 'Subfamilia não encontrada'}), 404

    except pyodbc.Error as e:
        print(e)
        return jsonify({'success': False, 'error': str(e)}), 500

@app.route('/delete_subfamilia', methods=['POST'])
def delete_subfamilia():
    data = request.get_json()
    delIdSubFamilia = data['delIdSubFamilia']
    delDescripcion = data['delDescripcion']
    delIdFamilia = data['delIdFamilia']
    delServicio = data['delServicio']    
    Activo = 1

    try:
        session = Session()
        Subfamilia = session.query(SubFamilia).filter(
            and_ (
                SubFamilia.IdSubFamilia == delIdSubFamilia,
                text("convert(varchar, Descripcion) = :desc").params(desc=delDescripcion), SubFamilia.IdSubFamilia == delIdSubFamilia,
                SubFamilia.Servicio == delServicio
            )
        ).first()
        if Subfamilia:
            Subfamilia.IdSubFamilia = delIdSubFamilia
            Subfamilia.Descripcion = delDescripcion
            Subfamilia.IdFamilia = delIdFamilia
            Subfamilia.Servicio = delServicio
            
            Subfamilia.Activo = Activo
            session.delete(Subfamilia)
            session.commit()
            session.close()

            return jsonify({'success': True}), 200
        
        return jsonify({'success': True}), 200



    except pyodbc.Error as e:
        print(e)
        return jsonify({'success': False, 'error': str(e)}), 500
